### PR TITLE
[SEARCH-391] Correctly handle negations in conjunction with nested queries.

### PR DESCRIPTION
**Pull request by gnusi**
### Scope & Purpose

Correctly handle negations in conjunction with nested queries.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 1]
```diff
@@ -1625,9 +1649,10 @@ Result fromInArray(irs::boolean_filter* filter, QueryContext const& ctx,
...
   if (filter) {
     filter = aql::NODE_TYPE_OPERATOR_BINARY_NIN == node.type
-                 ? &static_cast<irs::boolean_filter&>(
-                       filter->add<irs::Not>().filter<irs::Or>())
-                 : &static_cast<irs::boolean_filter&>(filter->add<irs::Or>());
+                 ? &static_cast<irs::boolean_filter&>(appendGroup<irs::Or>(
+                       filter->add<irs::Not>(), ctx.hasNestedFields))
+                 : &static_cast<irs::boolean_filter&>(
+                       appendGroup<irs::Or>(*filter, ctx.hasNestedFields));
     filter->boost(filterCtx.boost);
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1656]: NIT

Better to move condition to function parameter

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 2]
```diff
@@ -1779,9 +1805,10 @@ Result fromIn(irs::boolean_filter* filter, QueryContext const& ctx,
       }
 
       filter = aql::NODE_TYPE_OPERATOR_BINARY_NIN == node.type
-                   ? &static_cast<irs::boolean_filter&>(
-                         filter->add<irs::Not>().filter<irs::Or>())
-                   : &static_cast<irs::boolean_filter&>(filter->add<irs::Or>());
+                   ? &static_cast<irs::boolean_filter&>(appendGroup<irs::Or>(
+                         filter->add<irs::Not>(), ctx.hasNestedFields))
+                   : &static_cast<irs::boolean_filter&>(
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFilterFactory.cpp:1810]: same here

**Review by MBkkt — APPROVED** [review]

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 3]
```diff
@@ -908,10 +928,10 @@ Result fromGeoDistanceInterval(irs::boolean_filter* filter,
       return {TRI_ERROR_INTERNAL, "Malformed search/filter context"};
     }
 
-    auto& geo_filter =
-        (aql::NODE_TYPE_OPERATOR_BINARY_NE == node.cmp
-             ? filter->add<irs::Not>().filter<GeoDistanceFilter>()
-             : filter->add<GeoDistanceFilter>());
+    auto& geo_filter = (aql::NODE_TYPE_OPERATOR_BINARY_NE == node.cmp
+                            ? append<irs::Not>(*filter, ctx.hasNestedFields)
+                                  .filter<GeoDistanceFilter>()
```
Comment (Dronplane) [inline arangod/IResearch/IResearchFilterFactory.cpp:933]: Should we always use `append` for filter combining? As technically  any filter could use `all` inside

**Review by Dronplane — APPROVED** [review]
LGTM. Some minor comments

**Review by Dronplane — APPROVED** [review]
LGTM

**Review by MBkkt — APPROVED** [review]

**Pull request merged**